File: 2025-season6/6주차/최영권/DailyTemperatures.java
package LinkedList;

import java.util.Arrays;
import java.util.Stack;

/**
 * 시간복잡도 : temperatures 의 길이가 N일때, O(n)
 */
public class DailyTemperatures {
    public static void main(String[] args) {
        DailyTemperatures dailyTemperatures = new DailyTemperatures();
        int[] ints = dailyTemperatures.dailyTemperatures(new int[]{73, 74, 75, 71, 69, 72, 76, 73});
        System.out.println(Arrays.toString(ints));
    }

    public int[] dailyTemperatures(int[] temperatures) {
        int n = temperatures.length;
        int[] answer = new int[n];  // 기본값 0으로 초기화됨
        Stack<Integer> stack = new Stack<>();  // 인덱스를 저장

        for (int i = 0; i < n; i++) {
            // 현재 온도가 스택에 있는 날들의 온도보다 높으면
            while (!stack.isEmpty() && temperatures[stack.peek()] < temperatures[i]) {
                int prevIndex = stack.pop();
                answer[prevIndex] = i - prevIndex;  // 며칠 후인지 계산
            }

            // 현재 인덱스를 스택에 추가
            stack.push(i);
        }

        return answer;
    }
}

File: 2025-season6/6주차/남혜민/daily-temperatures.ts
// 스택으로 다음 더 따뜻한 날까지의 일수 계산
// 전체 시간 복잡도: O(n) - 각 인덱스는 최대 한 번 push/pop
// 전체 공간 복잡도: O(n) - 스택에 최대 n개의 인덱스 저장
function dailyTemperatures(temperatures: number[]): number[] {
    const n = temperatures.length;
    const result = new Array(n).fill(0); // O(n) - 결과 배열 초기화
    const stack: number[] = []; // 인덱스 저장하는 스택

    // 배열 순회 - O(n)
    for (let i = 0; i < n; i++) {
        // 현재 온도가 스택 top 인덱스의 온도보다 높으면
        // 각 인덱스는 최대 한 번만 pop됨 - 전체적으로 O(n)
        while (
            stack.length > 0 &&
            temperatures[i] > temperatures[stack[stack.length - 1]] // 온도 비교
        ) {
            const prevIndex = stack.pop()!; // 이전 인덱스 꺼내기
            result[prevIndex] = i - prevIndex; // 일수 차이 계산
        }
        stack.push(i); // 현재 인덱스 저장
    }

    return result;
}

console.log(dailyTemperatures([73, 74, 75, 71, 69, 72, 76, 73])); // [1,1,4,2,1,1,0,0]
